PSReadLine/ViRegister.cs

```diff
@@ -0,0 +1,239 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.PowerShell
+{
+    public partial class PSConsoleReadLine
+    {
+        /// <summary>
+        /// Used to report when the buffer is about to change
+        /// as part of a paste operation.
+        /// </summary>
+        public sealed class PasteEventArgs : EventArgs
+        {
+            /// <summary>
+            /// The text being pasted.
+            /// </summary>
+            public string Text { get; set; }
+            /// <summary>
+            /// The position in the buffer at
+            /// which the pasted text will be inserted.
+            /// </summary>
+            public int Position { get; set; }
+            /// <summary>
+            /// The recorded position in the buffer
+            /// from which the paste operation originates.
+            /// This is usually the same as Position, but
+            /// not always. For instance, when pasting a
+            /// linewise selection before the current line,
+            /// the Anchor is the cursor position, whereas
+            /// the Position is the beginning of the previous line.
+            /// </summary>
+            public int Anchor { get; set; }
+        }
+
+        /// <summary>
+        /// Represents a named register.
+        /// </summary>
+        public sealed class ViRegister
+        {
+            private string _text;
+            private bool _linewise;
+
+            /// <summary>
+            /// Raised when the text from this register is about
+            /// to be pasted to the buffer specified as part of
+            /// a call to Paste[After|Before].
+            /// </summary>
+            public event EventHandler<PasteEventArgs> OnInserting;

```

Comment:
It does not, no. I will make it internal.

Reply:
Well, I need some sort of callback mechanism between the `ViRegister` class and the Undo/Redo feature of the `PSReadLine` class. But it certainly does not need to be public.

If you prefer, I can change this into a method that registers a callback instead of an event.

Reply:
You have a single handler and it looks like it could be inlined. Do you have other callbacks in mind?

Reply:
I don’t have other callbacks in mind, no. But, I fail to understand what you are driving at.

I wanted to avoid a direct coupling between the `ViRegister` and the `PSReadLine` class. In essence, the `ViRegister` needs to call a method on the `PSReadLine` to undo a paste operation.

I could achieve this by passing a reference to the `PSReadLine`  `_singleton` object as an argument to the `ViRegister` constructor, but that is what I was trying to avoid.

The only two solutions I see are :
a) An event.
b) A method that accepts a callback.

Sorry about my misunderstanding.

Reply:
The `_singleton` object is pretty well baked in at this point, so a callback here felt like slight over-engineering.

Reply:
I have made the `_clipboard` member static and initialized both `_singleton` and `_clipboard` objects from a static constructor.